### ISSUE TITLE
HPUX - Unable to build with asm

### ISSUE THREAD

**Issue by RobinGeffroy** [openssl/openssl#17781]
Hello,

when I try to build OpenSSL on HPUX ia64 (itanium) with the native compiler (cc), I have to go with the "no-asm" option. Otherwise, the build fails.
I have the following error :
`ar: crypto/chacha/libcrypto-lib-chacha-ia64.o: No such file or directory`
I can see the compile line :  
`cc  -I. -Iinclude -Iproviders/common/include -Iproviders/implementations/include  -DAES_ASM -DGHASH_ASM -DOPENSSL_CPUID_OBJ -DPOLY1305_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM +Z -Ae +DD64 +Olit=all -z +O3 -DB_ENDIAN -DOPENSSL_PIC -D_REENTRANT -DOPENSSL_BUILDING_OPENSSL -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED -D_HPUX_ALT_XOPEN_SOCKET_API -DNDEBUG -DOPENSSL_TLS_SECURITY_LEVEL=0 -c -o crypto/chacha/libcrypto-lib-chacha-ia64.o crypto/chacha/chacha-ia64.S`

But no file is generated.
However, if I rename the file "chacha-ia64.S" to "chacha-ia64.s", the expected .o is generated.
It seems this compiler is rather picky when it comes to file extensions, and it *only* accepts .s...

Unfortunately, I have no idea how to fix it, since : 
  * I'm not familiar enough with HPUX cc compiler to know whether or not there is an option to handle .S instead of .s
  * I'm not familiar enough with OpenSSL build, I saw some files named .s and others .S, and don't know what the impact is

I can easily test any patch provided if needed, on a whole bunch of OSes as well. 

Note that :
  * after renaming chacha-ia64.S and poly1305-ia64.S to .s, the build was fine
  * using no-asm generate a core dump (BUS error) when trying to use AES_set_encrypt_key() (when trying to access key->round in aes_core.c, line 648 with tag openssl-3.0.1) : 
```
Program received signal SIGBUS, Bus error
  si_code: 1 - BUS_ADRALN - Invalid address alignment. Please refer to the following link that helps in handling unaligned data: http://docs.hp.com/en/7730/newhelp0610/pragmas.htm#pragma-
pack-ex3.
0x9fffffffbe53a730:0 in AES_set_decrypt_key () at crypto/aes/aes_core.c:648
648             key->rounds = 14;
(gdb) bt
#0  0x9fffffffbe53a730:0 in AES_set_decrypt_key () at crypto/aes/aes_core.c:648
#1  0x9fffffffbf5a8a00:0 in <caller>
```

**Comment by t8m** [comment]
Unfortunately there would have to be some additional step to generate the .s from .S if the compiler is cc.

**Comment by levitte** [comment]
So it seems that there is a running assumption that all Unixly C compilers take a `.S` file as assembler with C preprocessor directives, and a `.s` file as pure assembler.  HP appears not to agree with that "standard".

I found a statement in the BN build.info that appears to deal with this problem:

https://github.com/openssl/openssl/blob/828613c3e44ca1de6407a37d34de64c673ca61e9/crypto/bn/build.info#L37

https://github.com/openssl/openssl/blob/828613c3e44ca1de6407a37d34de64c673ca61e9/crypto/bn/build.info#L167

In other words, something similar should be used in the cases that hurt the HP compiler.  Generically speaking, it would look like this in the appropriate build.info files (with paths adapted to match):

From .pl to .s via .S:

```
SOURCE[../../libcrypto]=whatever.s

GENERATE[whatever.s]=whatever.S
GENERATE[whatever.S]=asm/whatever.pl
```

From .S to .s:

```
SOURCE[../../libcrypto]=whatever.s

GENERATE[whatever.s]=asm/whatever.S
```

**Comment by mattcaswell** [comment]
What does the `GENERATE` directive do to generate a .s file from a .S file? Is this something we should with the HP compiler only?

**Comment by rsbeckerca** [comment]
> What does the `GENERATE` directive do to generate a .s file from a .S file? Is this something we should with the HP compiler only?

I'm not sure what exactly c99 does on HPUX, but on HPNonStop, we do not have access to asm. It might be a similar situation.

**Comment by levitte** [comment]
> > What does the `GENERATE` directive do to generate a .s file from a .S file? Is this something we should with the HP compiler only?
> 
> I'm not sure what exactly `c99` does on HPUX, but on HPE NonStop, we do not have access to `asm`. It might be a similar situation. Assembler is not externalized for users.

Does that mean that NonStop targets expect to be configured with `no-asm`?

**Comment by levitte** [comment]
> > > > What does the `GENERATE` directive do to generate a .s file from a .S file? Is this something we should with the HP compiler only?
> > > 
> > > 
> > > I'm not sure what exactly `c99` does on HPUX, but on HPE NonStop, we do not have access to `asm`. It might be a similar situation. Assembler is not externalized for users.
> > 
> > 
> > Does that mean that NonStop targets expect to be configured with `no-asm`?
> 
> If there is an assumption that an assembler exists, then yes, definitely. It has not been an issue in past.

Never mind.  Getting assembler built requires that the attribute `asm_arch` be present in the relevant config target.  Since no NonStop includes that attribute, no asm will be compiled for them

**Comment by RobinGeffroy** [comment]
> What does the `GENERATE` directive do to generate a .s file from a .S file? Is this something we should with the HP compiler only?

I checked with gcc on HPUX and everything is fine. With the HP compiler, however, we have the issue.
I don't have the latest compiler update (mine is from.... 2008), and I asked the HP community about this, but no answer so far. I doubt HP fixed it in latest compiler releases, to be honest.

**Comment by t8m** [comment]
By this PR I mean https://github.com/openssl/openssl/pull/18263

**Comment by t8m** [comment]
Fixed by #18263
